firebase-firestore/src/main/java/com/google/firebase/firestore/local/MemoryRemoteDocumentCache.java

```diff
@@ -59,75 +57,67 @@ public void remove(DocumentKey key) {
     docs = docs.remove(key);
   }
 
-  @Nullable
   @Override
-  public MaybeDocument get(DocumentKey key) {
-    Pair<MaybeDocument, SnapshotVersion> entry = docs.get(key);
-    return entry != null ? entry.first : null;
+  public MutableDocument get(DocumentKey key) {
+    Pair<MutableDocument, SnapshotVersion> entry = docs.get(key);
+    return entry != null ? entry.first.clone() : new MutableDocument(key);
   }
 
   @Override
-  public Map<DocumentKey, MaybeDocument> getAll(Iterable<DocumentKey> keys) {
-    Map<DocumentKey, MaybeDocument> result = new HashMap<>();
-
+  public Map<DocumentKey, MutableDocument> getAll(Iterable<DocumentKey> keys) {
+    Map<DocumentKey, MutableDocument> result = new HashMap<>();
     for (DocumentKey key : keys) {
-      // Make sure each key has a corresponding entry, which is null in case the document is not
-      // found.
       result.put(key, get(key));
     }
-
     return result;
   }
 
   @Override
-  public ImmutableSortedMap<DocumentKey, Document> getAllDocumentsMatchingQuery(
+  public ImmutableSortedMap<DocumentKey, MutableDocument> getAllDocumentsMatchingQuery(
       Query query, SnapshotVersion sinceReadTime) {
     hardAssert(
         !query.isCollectionGroupQuery(),
         "CollectionGroup queries should be handled in LocalDocumentsView");
-    ImmutableSortedMap<DocumentKey, Document> result = emptyDocumentMap();
+    ImmutableSortedMap<DocumentKey, MutableDocument> result = emptyDocumentMap();
 
     // Documents are ordered by key, so we can use a prefix scan to narrow down the documents
     // we need to match the query against.
     ResourcePath queryPath = query.getPath();
     DocumentKey prefix = DocumentKey.fromPath(queryPath.append(""));
-    Iterator<Map.Entry<DocumentKey, Pair<MaybeDocument, SnapshotVersion>>> iterator =
+    Iterator<Map.Entry<DocumentKey, Pair<MutableDocument, SnapshotVersion>>> iterator =
         docs.iteratorFrom(prefix);
 
     while (iterator.hasNext()) {
-      Map.Entry<DocumentKey, Pair<MaybeDocument, SnapshotVersion>> entry = iterator.next();
+      Map.Entry<DocumentKey, Pair<MutableDocument, SnapshotVersion>> entry = iterator.next();
 
       DocumentKey key = entry.getKey();
       if (!queryPath.isPrefixOf(key.getPath())) {
         break;
       }
 
-      MaybeDocument maybeDoc = entry.getValue().first;
-      if (!(maybeDoc instanceof Document)) {
-        continue;
-      }
-
-      SnapshotVersion readTime = entry.getValue().second;
-      if (readTime.compareTo(sinceReadTime) <= 0) {
-        continue;
-      }
+      MutableDocument doc = entry.getValue().first;
```

Comment:
Personally, I prefer the "guard" style from the original code. The new code mixes an "if" statement and a guard, which is arguably more confusing. For readability, could you switch to exclusively use the "guard" style?

i.e.

```java
while (iterator.hasNext()) {
  Map.Entry<DocumentKey, Pair<MutableDocument, SnapshotVersion>> entry = iterator.next();

  DocumentKey key = entry.getKey();
  if (!queryPath.isPrefixOf(key.getPath())) {
    break;
  }

  MutableDocument doc = entry.getValue().first;
  if (! doc.isFoundDocument()) {
    continue;
  }
  
  SnapshotVersion readTime = entry.getValue().second;
  if (readTime.compareTo(sinceReadTime) <= 0) {
    continue;
  }

  if (!query.matches(doc)) {
    continue;
  }
  
  result = result.insert(doc.getKey(), doc.clone());
}
```